File: packages/engine/src/schemas/domain.ts
import { z } from "zod";

export const Season = z.enum(["spring", "summer", "fall", "winter", "year_round"]);
export type Season = z.infer<typeof Season>;

export const Severity = z.enum(["allergy", "intolerance", "preference"]);
export type Severity = z.infer<typeof Severity>;

/**
 * How a restricted person is handled for a given meal.
 * "split": cook normally, pull that person's portion before the restricted
 * ingredient is added (requires the ingredient to have added_at_step set).
 */
export const Handling = z.enum(["clear", "substitute", "split", "skip"]);
export type Handling = z.infer<typeof Handling>;

export const Lifecycle = z.enum(["to_try", "probation", "keeper", "cut"]);
export type Lifecycle = z.infer<typeof Lifecycle>;

export const StoreSection = z.enum([
  "produce",
  "meat_seafood",
  "dairy",
  "pantry",
  "frozen",
  "bakery",
  "other",
]);
export type StoreSection = z.infer<typeof StoreSection>;

export const Perishability = z.enum([
  "tender_herb", // cilantro, basil, mint, dill — the usual waste
  "hardy_herb", // thyme, rosemary, ginger, scallions — keep well
  "perishable", // fresh greens, sprouts, delicate produce
  "stable", // pantry, frozen, hardy produce
]);
export type Perishability = z.infer<typeof Perishability>;

export const CanonicalItem = z.object({
  id: z.string(),
  name: z.string(),
  aliases: z.array(z.string()).default([]),
  storeSection: StoreSection,
  perishability: Perishability,
  /** Allergen classes this item belongs to. */
  allergens: z.array(z.string()).default([]),
  /**
   * Negative assertions: classes this item is explicitly NOT in, even though
   * its name suggests otherwise (coconut milk → not dairy). Used to defend
   * against regressions if anyone ever adds name-based matching.
   */
  notAllergens: z.array(z.string()).default([]),
  typicalPriceCents: z.number().int().nonnegative().optional(),
});
export type CanonicalItem = z.infer<typeof CanonicalItem>;

export const RecipeIngredient = z.object({
  canonicalItemId: z.string().nullable(),
  /** Raw text as imported; kept for display and for unmatched items. */
  raw: z.string(),
  qty: z.number().positive().nullable(),
  unit: z.string().nullable(),
  prepNote: z.string().optional(),
  isOptional: z.boolean().default(false),
  /** 1-based step index at which this ingredient enters the dish; enables Split. */
  addedAtStep: z.number().int().positive().nullable().default(null),
});
export type RecipeIngredient = z.infer<typeof RecipeIngredient>;

export const RecipeStep = z.object({
  order: z.number().int().positive(),
  text: z.string(),
  durationMin: z.number().int().nonnegative().nullable().default(null),
  /** True for hands-off time (marinating, braising, resting). */
  unattended: z.boolean().default(false),
});
export type RecipeStep = z.infer<typeof RecipeStep>;

export const Recipe = z.object({
  id: z.string(),
  title: z.string(),
  serves: z.number().int().positive(),
  publishedTimeMin: z.number().int().positive().nullable().default(null),
  protein: z.string(),
  cuisine: z.string(),
  seasons: z.array(Season).min(1),
  equipment: z.array(z.string()).default([]),
  effort: z.enum(["easy", "moderate", "involved"]),
  ingredients: z.array(RecipeIngredient),
  steps: z.array(RecipeStep),
});
export type Recipe = z.infer<typeof Recipe>;

export const DietaryRestriction = z.object({
  personId: z.string(),
  /** Either an allergen class (e.g. "dairy") or a specific canonical item id. */
  allergenClass: z.string().nullable(),
  canonicalItemId: z.string().nullable(),
  severity: Severity,
  /** Free-text nuance, e.g. "hard aged cheeses OK; coconut milk is fine". */
  notes: z.string().optional(),
});
export type DietaryRestriction = z.infer<typeof DietaryRestriction>;

export const Person = z.object({
  id: z.string(),
  name: z.string(),
  isChild: z.boolean().default(false),
  /** Children on standing meals are excluded from dinner planning. */
  eatsPlannedDinners: z.boolean().default(true),
});
export type Person = z.infer<typeof Person>;

export const HouseholdRecipeState = z.object({
  recipeId: z.string(),
  lifecycle: Lifecycle,
  lastMadeAt: z.string().nullable().default(null), // ISO date
  timesMade: z.number().int().nonnegative().default(0),
  avgRating: z.number().min(1).max(5).nullable().default(null),
  /** Measured start-to-eating, from cook sessions. Never overwrites publishedTimeMin. */
  realTimeMin: z.number().int().positive().nullable().default(null),
});
export type HouseholdRecipeState = z.infer<typeof HouseholdRecipeState>;

export const PantryItem = z.object({
  canonicalItemId: z.string(),
  qty: z.number().nonnegative(),
  unit: z.string(),
  confidence: z.enum(["confirmed", "assumed"]),
});
export type PantryItem = z.infer<typeof PantryItem>;

export const StandingOrderLine = z.object({
  canonicalItemId: z.string().nullable(),
  raw: z.string(),
  qty: z.number().positive().nullable().default(null),
  unit: z.string().nullable().default(null),
  reason: z.string().optional(), // e.g. "kids' standing meals", "restock"
});
export type StandingOrderLine = z.infer<typeof StandingOrderLine>;

export const Household = z.object({
  id: z.string(),
  name: z.string(),
  /** Hemisphere + climate driver for seasonality; v1: month→season table by region. */
  region: z.enum(["northern", "southern"]).default("northern"),
  people: z.array(Person),
  restrictions: z.array(DietaryRestriction),
  equipment: z.array(z.string()).default([]),
  dinnersPerWeek: z.number().int().min(1).max(7).default(5),
  budgetCentsWeekly: z.number().int().positive().nullable().default(null),
  budgetEnabled: z.boolean().default(false),
});
export type Household = z.infer<typeof Household>;

File: service/mealplanner/schemas/domain.py
"""Domain models — a faithful port of packages/engine/src/schemas/domain.ts."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Season = Literal["spring", "summer", "fall", "winter", "year_round"]
Severity = Literal["allergy", "intolerance", "preference"]
# "split": cook normally, pull that person's portion before the restricted
# ingredient is added (requires the ingredient's added_at_step to be known).
Handling = Literal["clear", "substitute", "split", "skip"]
Lifecycle = Literal["to_try", "probation", "keeper", "cut"]
StoreSection = Literal[
    "produce", "meat_seafood", "dairy", "pantry", "frozen", "bakery", "other"
]
Perishability = Literal["tender_herb", "hardy_herb", "perishable", "stable"]
Effort = Literal["easy", "moderate", "involved"]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class CanonicalItem(StrictModel):
    id: str
    name: str
    aliases: list[str] = Field(default_factory=list)
    store_section: StoreSection
    perishability: Perishability
    allergens: list[str] = Field(default_factory=list)
    # Negative assertions: classes this item is explicitly NOT in even though
    # its name suggests otherwise (coconut milk -> not dairy). Documentation +
    # regression data; membership is decided only by `allergens`.
    not_allergens: list[str] = Field(default_factory=list)
    typical_price_cents: int | None = None


class RecipeIngredient(StrictModel):
    canonical_item_id: str | None = None
    raw: str
    qty: float | None = None
    unit: str | None = None
    prep_note: str | None = None
    is_optional: bool = False
    # Component this ingredient belongs to ("Sauce", "Marinade", "For the bowl")
    # when the source recipe groups them; None for ungrouped.
    group: str | None = None
    # 1-based step index at which this ingredient enters the dish; enables Split.
    added_at_step: int | None = None


class RecipeStep(StrictModel):
    order: int
    text: str
    duration_min: int | None = None
    # True for hands-off time (marinating, braising, resting).
    unattended: bool = False


class Recipe(StrictModel):
    id: str
    title: str
    serves: int = Field(gt=0)
    published_time_min: int | None = None
    protein: str
    cuisine: str
    seasons: list[Season] = Field(min_length=1)
    equipment: list[str] = Field(default_factory=list)
    effort: Effort = "moderate"
    ingredients: list[RecipeIngredient] = Field(default_factory=list)
    steps: list[RecipeStep] = Field(default_factory=list)


class DietaryRestriction(StrictModel):
    person_id: str
    # Either an allergen class (e.g. "dairy") or a specific canonical item id.
    allergen_class: str | None = None
    canonical_item_id: str | None = None
    severity: Severity
    notes: str | None = None


class Person(StrictModel):
    id: str
    name: str
    is_child: bool = False
    # Children on standing meals are excluded from dinner planning.
    eats_planned_dinners: bool = True


class HouseholdRecipeState(StrictModel):
    recipe_id: str
    lifecycle: Lifecycle = "to_try"
    last_made_at: str | None = None  # ISO date
    times_made: int = 0
    ratings: list[int] = Field(default_factory=list)
    notes: str | None = None
    # Measured start-to-eating. Never overwrites published_time_min.
    real_time_min: int | None = None


class PantryItem(StrictModel):
    canonical_item_id: str
    qty: float = Field(ge=0)
    unit: str
    confidence: Literal["confirmed", "assumed"] = "assumed"


class StandingOrderLine(StrictModel):
    canonical_item_id: str | None = None
    raw: str
    qty: float | None = None
    unit: str | None = None
    reason: str | None = None  # e.g. "kids' standing meals", "restock"


class Household(StrictModel):
    id: str
    name: str
    region: Literal["northern", "southern"] = "northern"
    people: list[Person] = Field(default_factory=list)
    restrictions: list[DietaryRestriction] = Field(default_factory=list)
    equipment: list[str] = Field(default_factory=list)
    # Proteins the household won't eat (e.g. "pork") — recipes with these
    # protein tags are filtered out deterministically, before planning.
    avoid_proteins: list[str] = Field(default_factory=list)
    # Canonical-item swaps applied to every grocery list (e.g. vegetable-oil
    # -> olive-oil): the banned item is never bought; its quantities aggregate
    # under the substitute instead.
    item_substitutions: dict[str, str] = Field(default_factory=dict)
    dinners_per_week: int = Field(default=5, ge=1, le=7)
    budget_cents_weekly: int | None = None
    budget_enabled: bool = False
